lib/rubygems/printable_uri.rb

```diff
@@ -0,0 +1,81 @@
+# frozen_string_literal: true
+
+require 'uri'
+require_relative 'uri_parser'
+
+class Gem::PrintableUri
+  def self.parse_uri(uri)
+    new(uri).parse_uri
+  end
+
+  def initialize(original_uri)
+    @credential_redacted = false
+    @original_uri = original_uri
+  end
+
+  def parse_uri
+    @original_uri = Gem::UriParser.parse_uri(@original_uri)
+    @uri = @original_uri.clone
+    redact_credential
+
+    self
+  end
+
+  def parsed_uri?
+    @uri.is_a? URI::Generic
```

Comment:
the `URI` could return `URI::HTTP` or `URI::HTTPS`, etc, when parsing a URI string but every class is inheriting the `URI::Generic`. 
I was thinking to rename this method to `parsable_uri?` but I wasn't sure.